### PR TITLE
Code overview tutorial

### PR DESCRIPTION
**Pull request by uri-granta**
Q: Should I also mention any of the following?

- VectorizedAcquisitionFunctionBuilder
- ReparametrizationSampler
- TrajectorySampler

### CODE REVIEW

**Comment by hstojic** [comment]
took a quick look and it nicely fills a gap, explaining important concepts in the toolbox, I'll take a more detailed look at a later point
also, `mk_multi_observer` is useful, I have used that snippet quite a few times already...

File: docs/notebooks/code_overview.pct.py [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+
+# %% [markdown]
+# ## Key types
+#
+# The following types represent the key concepts in Trieste. For a full listing of all the types in Trieste, see the API reference.
+
+# %% [markdown]
+# ### `Observer`
+#
+# The `Observer` type definition represents the black box objective function. Observers are functions that accept query points and return observations. Observations are either a single objective value that we wish to optimize, or a dictionary of mutiple tagged values that must be combined somehow, for example an objective and an inequality constraint. Objective values can be either single or multi-dimensional (see [multi-objective optimization](multi_objective_ehvi.ipynb)).
```
Comment (vpicheny) [inline docs/notebooks/code_overview.pct.py]: Observers return a dataset rather than observations, which is slightly unusual so we should state it.

Not sure how to write it, but it might be worth noting that this is very flexible in the sense that the returned dataset can be anything (not necessarily [query points, observations]).

File: docs/notebooks/code_overview.pct.py [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+
+# %% [markdown]
+# ### `ProbabilisticModel`
+#
+# The `ProbabilisticModel` protocol represents any probabilistic model used to model observations. Like for `Dataset`, observers with multiple observations are modelled by a dictionary o multiple tagged models.
+#
+# At it simplest, a `ProbabilisticModel` is anything that implements a `predict` and `sample` method. However, many algorithms in Trieste depend on models with additional features, which are represented by the various subclasses of `ProbabilisticModel`. The standard Bayesian optimizer uses `TrainableProbabilisticModel` models, which also implement `update` and `optimize` methods. Specific acuqisition functions may require other features, represented by classes like `SupportsPredictJoint` and `SupportsGetObservationNoise`. Since these are defined as protocols, it is possible to define and depend on the intersections of different model types (e.g. only support models that are both `SupportsPredictJoint` and `SupportsGetObservationNoise`).
```
Comment (vpicheny) [inline docs/notebooks/code_overview.pct.py]: It would be good to explain `update` and `optimize`. Maybe all in brackets?  

(...) which also implement `update` (updating the model structure when new data is added to the training set) and `optimize` (optimizing the model training loss) methods

also 
`SupportsPredictJoint` (ability to predict the joint distribution at several inputs x1, ... xp)
`SupportsGetObservationNoise` (ability to predict the observation noise variance)

File: docs/notebooks/code_overview.pct.py [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+#
+# Multiple models can also be combined into a single `ModelStack` model that combines their outputs for prediction and sampling. This can be useful when modelling multi-objective observations with independent models. There are also constructors like `TrainableModelStack` and `PredictJointModelStack` that combine specific types of model for use with code that requires that type, delegating the relevant methods where appropriate.
```
Comment (vpicheny) [inline docs/notebooks/code_overview.pct.py]: Would start with a slightly more general sentence about multiple outputs perhaps? e.g. something along the lines of

There are multiple ways to deal with multiple outputs: one can use a single multi-output model or a dictionary of models.

Maybe also add that `ModelStack` essentially emulates a single multi-output model from independent single-output models? and say it's useful for acquisitions that expect a multioutput model such as our multi-objective ones?

File: docs/notebooks/code_overview.pct.py [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+
+# %% [markdown]
+# ### `SearchSpace`
+#
+# The `SearchSpace` base class represents the domain over which the objective function is to be optimized. Spaces can currently be either continuous `Box` spaces, discrete `DiscreteSearchSpace` spaces, or a `TaggedProductSearchSpace` product of multipe spaces. All search spaces expose their dimensions, bounds and a sampler.
```
Comment (vpicheny) [inline docs/notebooks/code_overview.pct.py:34]: "over which the objective function is to be optimized"

mmmm I think that actually it's "over which the acquisition function is to be optimized"?? which in the case of trust regions is not the same. Possibly an unnecessary distinction.

File: docs/notebooks/code_overview.pct.py [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+# %% [markdown]
+# ### `AcquisitionRule`
+#
+# The `AcquisitionRule` base class represents a routine for selecting new query points during a Bayesian optimization loop (via an `acquire` method). It is generic on three types:
+#
+# * **ResultType**: the output of the rule, typically this is just tensors representing the query points. However, it can also be functions that accept some *acquisition state* and return the query points with a new state.
+# * **SearchSpaceType**: the type of the search space; any optimizer that the rule uses must support this.
+# * **ProbabilisticModelType**: the type of the models; any acquisition functions or samplers that the rule uses must support this.
+#
+# Exampes of rules include:
```
Comment (vpicheny) [inline docs/notebooks/code_overview.pct.py]: typo: examples

File: docs/notebooks/code_overview.pct.py [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+# 1. `AsynchronousOptimization` is similar to EGO but uses acquisition state to keep track of points whose observations are still pending.
+# 1. `DiscreteThompsonSampling` uses Thompson samplers rather than acquisition functions to select new query points.
+#
+
+# %% [markdown]
+# ### `AcquisitionFunction` and `AcquisitionFunctionBuilder`
+#
+# The `AcquisitionFunction` type definition represents any acquisition function: that is, a function that maps a set of query points to a single value that describes how useful it would be evaluate all these points together.
+#
+# The `AcquisitionFunctionBuilder` base class, meanwhile, represents something that builds and updates acquisition functions. At the start of the Bayesian optimization, the builder's `prepare_acquisition_function` method is called by the acquisition rule to create an acquisition function from the current observations and probabilistic models. For efficiency, most builders also define an `update_acquisition_function` method for updating the function using the updated observations and models. (The ones that don't instead generate a new acquisition function when necessary.)
```
Comment (vpicheny) [inline docs/notebooks/code_overview.pct.py]: maybe give a little bit more detail about "for efficiency"? e.g. for someone familiar with BO but not tensorflow, this is probably very puzzling (it's still very puzzling for me).

File: docs/notebooks/code_overview.pct.py [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+#
+# Acquisition functions that support only one probabilistic model are more easily defined using the `SingleModelAcquisitionBuilder` convenience class.
+#
+# Acquisition functions that are suitable for greedily building batches of points can be defined using `GreedyAcquisitionFunctionBuilder` (or `SingleModelGreedyAcquisitionBuilder`) using a `prepare_acquisition_function` method that also accepts the points already chosen to be in the current batch.
+
+# %% [markdown]
+# ### `AcquisitionOptimizer`
+#
+# The `AcquisitionOptimizer` type definition represents an optimizer function that maximizes an acquisition function over a search space. Trieste provides a `generate_continuous_optimizer` function for generating gradient-based optimizers for continuous spaces and `automatic_optimizer_selector` for quickly selecting an appropriate optimizer.
```
Comment (vpicheny) [inline docs/notebooks/code_overview.pct.py]: worth telling here that we can also deal with discrete and hybrid cases?

File: docs/notebooks/code_overview.pct.py [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+
+# %% [markdown]
+# ### `BayesianOptimizer` and `AskTellOptimizer`
+#
+# The `BayesianOptimizer` and `AskTellOptimizer` classes are the two Bayesian optimization interfaces provided by Trieste.
+#
+# `BayesianOptimizer` exposes an `optimize` method for running a Bayesian optimization loop with given initial datasets and models, and a given number of steps (and an optional early stop callback).
```
Comment (vpicheny) [inline docs/notebooks/code_overview.pct.py:73]: add that there is some monitoring as well?

File: docs/notebooks/code_overview.pct.py [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+#
+# `AskTellOptimizer` provides greater control over the loop, by providing separate `ask` and `tell` steps for suggesting query points and updating the models with new observations.
```
Comment (vpicheny) [inline docs/notebooks/code_overview.pct.py]: link to the notebook?

File: docs/notebooks/code_overview.pct.py [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+import tensorflow as tf
+from trieste.types import TensorType
+
+# %% [markdown]
+# This section explains how to define new observers, model types and acqusition functions.
+
+# %% [markdown]
+# ### New observers
+#
+# Defining an observer with a single observation is as simple as defining a function that returns that observation:
```
Comment (vpicheny) [inline docs/notebooks/code_overview.pct.py]: an dataset

File: docs/notebooks/code_overview.pct.py [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+)
+multiobserver(tf.constant([[0, 1], [1, 1]], dtype=tf.float64))
+
+# %% [markdown]
+# ### New probabilistic model types
+#
+# Defining a new probabilistic model type simply involves writing a class that implements all the relevant methods (at the very least `predict` and `sample`). For clarity, it is best to also explicitly inherit from the supported feature protocols.
+
+# %%
+from trieste.models.interfaces import (
```
Comment (vpicheny) [inline docs/notebooks/code_overview.pct.py:134]: trajectory samplers are advanced methods. Possibly doesn't matter, otherwise use for example a more trivial feature, eg `SupportGetObservationNoise?

File: docs/notebooks/code_overview.pct.py [hunk 12]
```diff
@@ -0,0 +1,238 @@
...
+    TrainableProbabilisticModel,
+    HasReparamSampler,
+    HasTrajectorySampler,
+    ReparametrizationSampler,
+)
+
+
+class GizmoModel(
```
Comment (vpicheny) [inline docs/notebooks/code_overview.pct.py:142]: I'm puzzled, we don't specify a vanilla sample method here.

File: docs/notebooks/code_overview.pct.py [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+    def reparam_sampler(
+        self, num_samples: int
+    ) -> ReparametrizationSampler[GizmoModel]:
+        ...
+
+    ...
+
+
+# %% [markdown]
+# If the new model type has an additional feature on which you'd like to depend, e.g. in a new acquisition function, then you can define that feature as a protocol. Marking it runtime_checkable will alow you to check for the feature at runtime too.
```
Comment (vpicheny) [inline docs/notebooks/code_overview.pct.py]: I don't know why I would want or not to check it at runtime. Add a clue?

File: docs/notebooks/code_overview.pct.py [hunk 14]
```diff
@@ -0,0 +1,238 @@
...
+
+# %%
+from trieste.models.interfaces import ProbabilisticModel
+from typing_extensions import Protocol, runtime_checkable
```
Comment (vpicheny) [inline docs/notebooks/code_overview.pct.py:165]: maybe add something specific about `Protocol`?

File: docs/notebooks/code_overview.pct.py [hunk 15]
```diff
@@ -0,0 +1,238 @@
...
+):
+    """A model that supports both gizmo, reparam_sampler and trajectory_sampler."""
+
+    pass
+
+
+# %% [markdown]
+# ### New acquisition function builders
+#
+# To define a new acquisition function builder, you simply need to define a class with a `prepare_acquisition_function` method that returns an `AcquisitionFunction`. If the acquisition function depends on just one model/dataset (as is often the case) then you can define it as a `SingleModelAcquisitionBuilder`; if it depends on more than one (e.g. both an objective and a constraint) then you must define it as a `ModelAcquisitionBuilder` instead. You can also specify (in brackets) the type of probabilistic models that the acquisition function supports: e.g. a `SingleModelAcquisitionBuilder[HasReparamSampler]` only supports models with a reparatemtrization sampler. This allows the type checker to warn you if you try to use an incompatible model type.
```
Comment (vpicheny) [inline docs/notebooks/code_overview.pct.py]: typo reparatemtrization

File: docs/notebooks/code_overview.pct.py [hunk 16]
```diff
@@ -0,0 +1,238 @@
...
+#
+
+# %% [markdown]
+# ### `AcquisitionFunction` and `AcquisitionFunctionBuilder`
+#
+# The `AcquisitionFunction` type definition represents any acquisition function: that is, a function that maps a set of query points to a single value that describes how useful it would be evaluate all these points together.
+#
+# The `AcquisitionFunctionBuilder` base class, meanwhile, represents something that builds and updates acquisition functions. At the start of the Bayesian optimization, the builder's `prepare_acquisition_function` method is called by the acquisition rule to create an acquisition function from the current observations and probabilistic models. For efficiency, most builders also define an `update_acquisition_function` method for updating the function using the updated observations and models. (The ones that don't instead generate a new acquisition function when necessary.)
+#
+# Acquisition functions that support only one probabilistic model are more easily defined using the `SingleModelAcquisitionBuilder` convenience class.
```
Comment (vpicheny) [inline docs/notebooks/code_overview.pct.py]: could be good to give a bit more detail about this? why it is convenient, what it changes...

File: docs/notebooks/code_overview.pct.py [hunk 17]
```diff
@@ -0,0 +1,238 @@
...
+
+# %%
+from trieste.acquisition import (
+    AcquisitionFunction,
+    SingleModelAcquisitionBuilder,
+)
+from trieste.data import Dataset
+
+
+class ProbabilityOfValidity(SingleModelAcquisitionBuilder[ProbabilisticModel]):
```
Comment (vpicheny) [inline docs/notebooks/code_overview.pct.py:203]: is it worth having a multi-model example as well? (not sure)

File: docs/notebooks/code_overview.pct.py [hunk 18]
```diff
@@ -0,0 +1,238 @@
...
+    ) -> AcquisitionFunction:
+        def acquisition(at: TensorType) -> TensorType:
+            mean, _ = model.predict_y(tf.squeeze(at, -2))
+            return mean
+
+        return acquisition
+
+
+# %% [markdown]
+# For efficiency, it usually makes sense to compile the generated acquisition function into a TensorFlow graph using `tf.function`. Furthermore, to avoid generating (and compiling) a new acquisition function on each Bayesian optimization loop, you can define an `update_acquisition_function` method that can instead update the previously generated acquisition function using the new models and data. This may involve updating the acquisition function's internal state (which you should store in `tf.Variable`s), though if the function has no internal state then it is suficient to simply return the old function unchanged.
```
Comment (vpicheny) [inline docs/notebooks/code_overview.pct.py:215]: this text should go to the top

File: docs/notebooks/code_overview.pct.py [hunk 19]
```diff
@@ -0,0 +1,238 @@
...
+
+# %%
+class ProbabilityOfValidity2(SingleModelAcquisitionBuilder[ProbabilisticModel]):
```
Comment (vpicheny) [inline docs/notebooks/code_overview.pct.py:218]: maybe more useful to have one example with a state?

**Review by vpicheny — COMMENTED** [review]
that's a great addition to the doc!

**Comment by vpicheny** [comment]
> Q: Should I also mention any of the following?
> 
>     * VectorizedAcquisitionFunctionBuilder
> 
>     * ReparametrizationSampler
> 
>     * TrajectorySampler

I don't think that it's documented anywhere... which is a problem. But not sure if this is the right place if we want to focus on the main items.
